Tutorials in ReadTheDocs documentation

VeckoTheGecko opened the pull request:
This PR works to address discussion from #1353 looking to add notebooks to the RTD documentation.

- [x] Move notebooks to `docs/examples` folder
- [x] Deprecate `parcels_get_examples`, and introduce new example data loading. This makes notebooks standalone, with data being downloaded and cached as required. Migrate all notebooks/tests to use new data loading.
- [x] Add tutorial gallery on RTD
- [x] Add download button to download all parcels examples notebooks as ZIP
- [x] Update notebook titles and update notebook H1 headings to match
- [x] Cleanup
  - Clean up unsupported markdown in notebooks
  - Fixing `Hx` headings

VeckoTheGecko commented:
Also the following notebooks aren't linked on the OceanParcels website:
* `documentation_LargeRunsOutput.ipynb`
* `documentation_advanced_zarr.ipynb`
* `documentation_homepage_animation.ipynb`
* `tutorial_Agulhasparticles.ipynb`
* `tutorial_analyticaladvection.ipynb`

VeckoTheGecko commented:
Let me know if you want me to explore other visualisation options that aren't a list of tutorial links (e.g. cards like before).

erikvansebille commented:
Thanks for initiating this, @VeckoTheGecko! I have a few very busy days ahead so not much time to look into this now, but will try to make some time at the end of the week. More soon!

VeckoTheGecko commented:
All good. I'm happy to take care of fixing the formatting stuff, and the notebook titles if that makes things easier.

For the updated titles for notebooks, if you give me these in a dictionary format as I'll be programatically updating the notebook titles. *See below for existing current notebook titles on the website and in the notebook itself.*

<details><summary>Notebook titles currently</summary>
<p>

```py
website_notebook_titles = {
    # Overview
    "tutorial_parcels_structure.ipynb": "General Parcels structure",
    "parcels_tutorial.ipynb": "Simple Parcels tutorial",
    "tutorial_output.ipynb": "Output tutorial",

    # Tutorials on setting up FieldSets
    "tutorial_periodic_boundaries.ipynb": "Periodic boundaries tutorial",
    "tutorial_interpolation.ipynb": "Interpolation tutorial",
    "tutorial_unitconverters.ipynb": "UnitConverter tutorial",
    "tutorial_timestamps.ipynb": "TimeStamps tutorial",
    "documentation_indexing.ipynb": "Grid indexing documentation",
    "tutorial_nemo_curvilinear.ipynb": "Curvilinear grid tutorial",
    "tutorial_nemo_3D.ipynb": "Nemo 3D tutorial",
    "tutorial_SummedFields.ipynb": "SummedFields tutorial",
    "tutorial_NestedFields.ipynb": "NestedFields tutorial",
    "tutorial_timevaryingdepthdimensions.ipynb": "Time-varying depths tutorial",

    # Tutorials on creating ParticleSets
    "tutorial_jit_vs_scipy.ipynb": "JIT Particles and Scipy particles",
    "tutorial_delaystart.ipynb": "Delay start tutorial",

    # Tutorials on writing kernels to be executed on each particle
    "tutorial_diffusion.ipynb": "Diffusion tutorial",
    "tutorial_sampling.ipynb": "Field Sampling tutorial",
    "tutorial_particle_field_interaction.ipynb": "Particle-Field interaction tutorial",
    "tutorial_interaction.ipynb": "Particle-Particle interaction tutorial",
    "tutorial_Argofloats.ipynb": "Argo Kernel tutorial",
    
    # Other tutorials
    "documentation_MPI.ipynb": "Parallelisation documentation",
    "documentation_stuck_particles.ipynb": "Stuck particle documentation",
    "documentation_unstuck_Agrid.ipynb": "Preventing stuck particles documentation",
    "tutorial_plotting.ipynb": "Plotting tutorial",

    # Not on website
    "documentation_LargeRunsOutput.ipynb": None,
    "documentation_advanced_zarr.ipynb": None,
    "documentation_homepage_animation.ipynb": None,
    "tutorial_Agulhasparticles.ipynb": None,
    "tutorial_analyticaladvection.ipynb": None,
}

current_notebook_titles = {
    # Overview
    "tutorial_parcels_structure.ipynb": "Getting started with Parcels: general structure",
    "parcels_tutorial.ipynb": "Parcels Tutorial",
    "tutorial_output.ipynb": "Tutorial on how to analyse Parcels output",

    # Tutorials on setting up FieldSets
    "tutorial_periodic_boundaries.ipynb": "Tutorial on how to implement periodic boundaries",
    "tutorial_interpolation.ipynb": "Tutorial on Parcels tracer interpolation methods",
    "tutorial_unitconverters.ipynb": "Tutorial on Units and UnitConverters in Parcels",
    "tutorial_timestamps.ipynb": "Tutorial on how to use `timestaps` in Field construction",
    "documentation_indexing.ipynb": "Documentation on dimensions and indexing in Parcels",
    "tutorial_nemo_curvilinear.ipynb": "Tutorial on how to use Parcels on NEMO curvilinear grids",
    "tutorial_nemo_3D.ipynb": "Tutorial on how to use Parcels on 3D C grids",
    "tutorial_SummedFields.ipynb": "Tutorial on how to combine different Fields for advection into a `SummedField` object",
    "tutorial_NestedFields.ipynb": "Tutorial on how to combine different Fields into a `NestedField` object",
    "tutorial_timevaryingdepthdimensions.ipynb": "Tutorial on how to use S-grids with time-evolving depth dimensions",

    # Tutorials on creating ParticleSets
    "tutorial_jit_vs_scipy.ipynb": "Tutorial on JIT versus Scipy execution within Parcels",
    "tutorial_delaystart.ipynb": "Tutorial on how to 'delay' the start of particle advection",

    # Tutorials on writing kernels to be executed on each particle
    "tutorial_diffusion.ipynb": "Tutorial: advection-diffusion kernels in Parcels",
    "tutorial_sampling.ipynb": "Field sampling tutorial",
    "tutorial_particle_field_interaction.ipynb": "Particle-Field interaction example",
    "tutorial_interaction.ipynb": "Particle-particle interaction example",
    "tutorial_Argofloats.ipynb": "Tutorial on how to simulate an Argo float in Parcels",

    # Other tutorials
    "documentation_MPI.ipynb": "Tutorial on Parallel Parcels with MPI and `dask` field-chunking",
    "documentation_stuck_particles.ipynb": "Documentation of particles 'stuck' on boundaries",
    "documentation_unstuck_Agrid.ipynb": "Tutorial on implementing boundary conditions in an A grid",
    "tutorial_plotting.ipynb": "Parcels plotting methods",

    # Not on website
    "documentation_LargeRunsOutput.ipynb": "Combining large ocean-parcels datasets and choosing an optimal chunking.",
    "documentation_advanced_zarr.ipynb": "Tutorial on Advanced Zarr Output",
    "documentation_homepage_animation.ipynb": None,
    "tutorial_Agulhasparticles.ipynb": "Tutorial showing how to create Parcels in Agulhas animated gif",
    "tutorial_analyticaladvection.ipynb": "Tutorial on the Analytical Advection kernel in Parcels",
}

```

</p>
</details>

VeckoTheGecko commented:
[Compacted diff](https://www.diffchecker.com/ACihcgFr) for 2359747 (for easy oversight) 😄

VeckoTheGecko commented:
### Data download refactor

- Managed to refactor the data downloading.
  - I used a similar approach to that used by `seaborn.load_dataset()`, which also implements caching.
- Updated example scripts and notebooks to use new data downloading approach
  - Still some integration tests to fix (not sure why they're failing, but don't have time to patch now)
  - Notebooks may need some rewording (eg. "We use the data in the `MovingEddies_data` folder" -> "The `MovingEddies_data` example dataset is downloaded")
  - How can the notebooks be worded such that the data downloading (/path insertion into the filename) isn't confusing to those new to Python?

### Download button
- Added button to download zip of tutorial notebooks in fc7d5e3. Button is in tutorials page, but can be surfaced as a large button on the docs homepage

With the example data refactor, notebooks are 100% self-contained.

### Broken anchor tags

Automatically finding broken anchor tags isn't trivial, so its easier to do manually. I did manage to make a list of all anchor tags, so that should help with a targeted search. If the anchor tag works in the notebook, it should work in the HTML when compiled.

```
                           documentation_MPI.ipynb : Anchor link found `#future-developments-load-balancing`
                  tutorial_parcels_structure.ipynb : Anchor link found `#1.-FieldSet`
                  tutorial_parcels_structure.ipynb : Anchor link found `#2.-ParticleSet`
                  tutorial_parcels_structure.ipynb : Anchor link found `#3.-Kernels`
                  tutorial_parcels_structure.ipynb : Anchor link found `#4.-Execution-and-Output`
                  tutorial_parcels_structure.ipynb : Anchor link found `#For-more-advanced-tutorials-on-writing-custom-kernels-that-work-on-custom-particles:`
                        tutorial_interaction.ipynb : Anchor link found `#mutator_function`
               documentation_stuck_particles.ipynb : Anchor link found `#Introduction`
               documentation_stuck_particles.ipynb : Anchor link found `#1.-A-grids`
               documentation_stuck_particles.ipynb : Anchor link found `#2.-C-grids`
               documentation_stuck_particles.ipynb : Anchor link found `#3.-B-grids`
               documentation_stuck_particles.ipynb : Anchor link found `#4.-Diffusion`
               documentation_stuck_particles.ipynb : Anchor link found `#C-grids`
               documentation_stuck_particles.ipynb : Anchor link found `#A-grids`
                            parcels_tutorial.ipynb : Anchor link found `#Running-particles-in-an-idealised-field`
                            parcels_tutorial.ipynb : Anchor link found `#Running-particles-in-backward-time`
                            parcels_tutorial.ipynb : Anchor link found `#Adding-a-custom-behaviour-kernel`
                            parcels_tutorial.ipynb : Anchor link found `#Reading-in-data-from-arbritrary-NetCDF-files`
                            parcels_tutorial.ipynb : Anchor link found `#Sampling-a-Field-with-Particles`
                            parcels_tutorial.ipynb : Anchor link found `#A-second-example-kernel:-calculating-distance-travelled`
                             tutorial_output.ipynb : Anchor link found `#Reading-the-output-file`
                             tutorial_output.ipynb : Anchor link found `#Trajectory-data-structure`
                             tutorial_output.ipynb : Anchor link found `#Analysis`
                             tutorial_output.ipynb : Anchor link found `#Plotting`
                             tutorial_output.ipynb : Anchor link found `#Animations`
                             tutorial_output.ipynb : Anchor link found `#Trajectory-data-structure`
                             tutorial_output.ipynb : Anchor link found `#Conditional-selection`
                           tutorial_sampling.ipynb : Anchor link found `#Basic-sampling`
                           tutorial_sampling.ipynb : Anchor link found `#Sampling-velocity-fields`
                           tutorial_sampling.ipynb : Anchor link found `#Sampling-initial-values`
                           tutorial_sampling.ipynb : Anchor link found `#Sampling-with-repeatdt`
                           tutorial_sampling.ipynb : Anchor link found `#repeatdt_error`
                 documentation_unstuck_Agrid.ipynb : Anchor link found `#1.-Particle-deletion`
                 documentation_unstuck_Agrid.ipynb : Anchor link found `#2.-Displacement`
                 documentation_unstuck_Agrid.ipynb : Anchor link found `#3.-Slip-boundary-conditions`

```

VeckoTheGecko commented:
Waiting on green for integration tests. I think thats it for this PR, is there anything else that I missed @erikvansebille ?

I'll leave working of tutorials and final checks up to you if thats alright (might be worth adding a section in an early intro tutorial mentioning what the `download_example_dataset()` function does so that users don't get confused, and calling `list_example_datasets()`).

erikvansebille commented on .readthedocs.yaml:
```diff
@@ -3,6 +3,9 @@ build:
   os: ubuntu-22.04
   tools:
     python: mambaforge-4.10
+  # jobs:
+  #   pre_build:
+  #     - sphinx-build -b linkcheck docs/ _build/linkcheck
```
Is this change needed, since it's commented out? An old debug statement?

erikvansebille commented on docs/conf.py:
Should we also change all the links to other notebooks in the tutorials and documentations to relative paths, so that readers stay within the docs and are not directed to nbviewer? Is there a way that changing these links (i.e. removing the `https://nbviewer.org/github/OceanParcels/parcels/blob/master/parcels/examples/` parts) can be scripted automatically?

VeckoTheGecko replied:
Done now in 546943a3bf4d62fc10bd1ee57319dc7072b3c0cd

VeckoTheGecko replied:
Done now in 546943a3bf4d62fc10bd1ee57319dc7072b3c0cd

erikvansebille commented on docs/documentation.rst:
```diff
@@ -0,0 +1,70 @@
+   examples/tutorial_diffusion.ipynb
+   examples/tutorial_sampling.ipynb
+   examples/tutorial_particle_field_interaction.ipynb
+   examples/tutorial_interaction.ipynb
```
This one still needs a thumbnail; the animation is on my other computer. Will upload it later

erikvansebille commented on docs/documentation.rst:
```diff
@@ -0,0 +1,70 @@
+   examples/documentation_MPI.ipynb
+   examples/documentation_stuck_particles.ipynb
+   examples/documentation_unstuck_Agrid.ipynb
+   examples/documentation_LargeRunsOutput.ipynb
```
This one also still needs a thumbnail. I don't really have an idea what could work here. Any suggestions?

VeckoTheGecko replied:
https://upload.wikimedia.org/wikipedia/commons/7/70/Oxygen480-devices-drive-harddisk.svg ?
![](https://upload.wikimedia.org/wikipedia/commons/7/70/Oxygen480-devices-drive-harddisk.svg)

erikvansebille commented on parcels/application_kernels/advectiondiffusion.py:
```diff
@@ -1,6 +1,6 @@
 """Collection of pre-built advection-diffusion kernels.
 
-See `this tutorial <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_diffusion.ipynb>`__ for a detailed explanation.
+See `this tutorial <../examples/tutorial_diffusion.ipynb>`__ for a detailed explanation.
```
Would this link work, given this file is not in the docs? Better to change to https://docs.oceanparcels.org/en/latest/examples/tutorial_diffusion.html (I think that'll be the correct link)?

erikvansebille commented on parcels/compilation/codegenerator.py:
```diff
@@ -624,7 +624,7 @@ def visit_Call(self, node):
                         node.ccode = rhs
             except:
-                raise RuntimeError("Error in converting Kernel to C. See https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_parcels_structure.ipynb#3.-Kernel-execution for hints and tips")
+                raise RuntimeError("Error in converting Kernel to C. See https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/docs/examples/tutorial_parcels_structure.ipynb#3.-Kernel-execution for hints and tips")
```
Link to the docs-version (https://docs.oceanparcels.org/en/latest/examples/tutorial_parcels_structure.ipynb#3.-Kernel-execution)?

erikvansebille commented on parcels/external/appdirs.py:
Why does this file need to be in Parcels? Can't it be installed as a package?

VeckoTheGecko replied:
Good call, it doesn't. I was mainly following the implementation from `seaborn, but using [`platformdirs`](https://anaconda.org/conda-forge/platformdirs) means we don't have to look at that code again.

Fixed

erikvansebille commented on parcels/field.py:
```diff
@@ -113,9 +113,9 @@ class Field:
     For usage examples see the following tutorials:
 
-    * `Nested Fields <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_NestedFields.ipynb>`__
+    * `Nested Fields <../examples/tutorial_NestedFields.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -113,9 +113,9 @@ class Field:
+    * `Nested Fields <../examples/tutorial_NestedFields.ipynb>`__
 
-    * `Summed Fields <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_SummedFields.ipynb>`__
+    * `Summed Fields <../examples/tutorial_SummedFields.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -348,7 +348,7 @@ def from_netcdf(cls, filenames, variable, dimensions, indices=None, grid=None,
         For usage examples see the following tutorial:
 
-        * `Timestamps <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_timestamps.ipynb>`__
+        * `Timestamps <../examples/tutorial_timestamps.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -638,7 +638,7 @@ def set_scaling_factor(self, factor):
         For usage examples see the following tutorial:
 
-        * `Unit converters <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_unitconverters.ipynb>`__
+        * `Unit converters <../examples/tutorial_unitconverters.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -651,7 +651,7 @@ def set_depth_from_field(self, field):
         Notes
         -----
-        See `this tutorial <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_timevaryingdepthdimensions.ipynb>`__
+        See `this tutorial <../examples/tutorial_timevaryingdepthdimensions.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -1351,7 +1351,7 @@ def add_periodic_halo(self, zonal, meridional, halosize=5, data=None):
         Before adding a periodic halo to the Field, it has to be added to the Grid on which the Field depends
 
-        See `this tutorial <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_periodic_boundaries.ipynb>`__
+        See `this tutorial <../examples/tutorial_periodic_boundaries.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -1972,7 +1972,7 @@ class SummedField(list):
     Examples
     --------
-    See `here <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_SummedFields.ipynb>`__
+    See `here <../examples/tutorial_SummedFields.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/field.py:
```diff
@@ -2053,7 +2053,7 @@ class NestedField(list):
     Examples
     --------
-    See `here <https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_NestedFields.ipynb>`__
+    See `here <../examples/tutorial_NestedFields.ipynb>`__
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/tools/converters.py:
```diff
@@ -268,5 +268,5 @@ def convert_xarray_time_units(ds, time):
                            'try using the timestamps keyword in the construction of your Field. '
                            'See also the tutorial at https://nbviewer.jupyter.org/github/OceanParcels/'
-                           'parcels/blob/master/parcels/examples/tutorial_timestamps.ipynb')
+                           'parcels/blob/master/docs/examples/tutorial_timestamps.ipynb')
```
Here too change to the docs.oceanparcels.org version of the page?

erikvansebille commented on parcels/utils.py:
Rename this file to a more explanatory name (e.g. `manage_example_dataset.py`)?

VeckoTheGecko replied:
I think these functions definitely fall into the utils category, and I think it would be good to keep this in case parcels would like to later introduce other utility functions.

erikvansebille commented on tests/test_examples.py:
```diff
@@ -1,32 +1,26 @@
-def test_peninsula():
-    from parcels.examples.example_peninsula import main
-    main(['jit'])
+# TODO: Broaden scope to all `.py` scripts in `example_folder`?
```
Indeed, why not add them all? (see below)
```suggestion
```

erikvansebille commented on tests/test_utils.py:
Also change the name of this file if changing the name of the utils.py file above

erikvansebille reviewed (commented):
What a Herculean task, @VeckoTheGecko! Thanks so much for all these changes, really makes the documentation much nicer to have the tutorials directly on docs.oceanparcels.org. 

A few comments lister here

VeckoTheGecko commented on tests/test_examples.py:
```diff
@@ -1,32 +1,26 @@
+# TODO: Broaden scope to all `.py` scripts in `example_folder`?
+example_folder = (Path(__file__).parent / "../docs/examples").resolve()
+example_fnames = [
+    "example_decaying_moving_eddy.py",
+    "example_moving_eddies.py",
+    # "example_nemo_curvilinear.py", #! Imports cartopy, which is not installed in CI
+    "example_peninsula.py",
+    "example_radial_rotation.py",
+    "example_stommel.py",
```
Done 7bafe84db3f8caa4dc8cb16dfcc1974fee543efb. Updated script finding to eager finding and testing `examples/*.py` (rather than hardcoded files). An FYI, most of these extra examples don't define a `main()` function, so the test case isn't really useful here (although they do define pytests).

VeckoTheGecko commented:
@erikvansebille could you add `platformdirs` to the forge recipe? I have no idea how that works

VeckoTheGecko commented:
Sphinx build process flagged the broken heading anchors. I've now corrected them

erikvansebille commented:
> @erikvansebille could you add `platformdirs` to the forge recipe? I have no idea how that works

Done in https://github.com/conda-forge/parcels-feedstock/pull/94
